From 5b5e8bb29ca41694b3dc2694779613f632f13ce3 Mon Sep 17 00:00:00 2001
From: IAP Desktop Build <iap-desktop+build@google.com>
Date: Fri, 22 Jul 2022 12:14:23 +1000
Subject: [PATCH 11/11] Expose properties for back/foreground color

Apply values immediately by updating defaults
(default/null attribute and the current cursor.
---
 VtNetCore.Unit.Tests/VirtualTerminal.cs       | 28 +++++++++++++++++
 .../VirtualTerminalController.cs              | 30 +++++++++++++++++++
 2 files changed, 58 insertions(+)

diff --git a/VtNetCore.Unit.Tests/VirtualTerminal.cs b/VtNetCore.Unit.Tests/VirtualTerminal.cs
index 39daaf1..3b93f38 100644
--- a/VtNetCore.Unit.Tests/VirtualTerminal.cs
+++ b/VtNetCore.Unit.Tests/VirtualTerminal.cs
@@ -565,5 +565,33 @@ namespace VtNetCore.Unit.Tests
             Assert.Equal(attributes.BackgroundColor, controller.CursorState.Attributes.BackgroundColor);
             Assert.Equal(VtNetCore.VirtualTerminal.Enums.ETerminalColor.Black, controller.CursorState.Attributes.ForegroundColor);
         }
+
+        [Fact]
+        public void ChangingForegroundColorAppliesToCursor()
+        {
+            var red = new TerminalColor() { Red = 255 };
+
+            var controller = new VirtualTerminalController(
+                TerminalAttribute.CreateDefault());
+
+            controller.ForegroundColor = red;
+
+            Assert.Equal(red, controller.DefaultAttribute.ForegroundRgb);
+            Assert.Equal(red, controller.CursorState.Attributes.ForegroundRgb);
+        }
+
+        [Fact]
+        public void ChangingBackgroundColorAppliesToCursor()
+        {
+            var red = new TerminalColor() { Red = 255 };
+
+            var controller = new VirtualTerminalController(
+                TerminalAttribute.CreateDefault());
+
+            controller.BackgroundColor = red;
+
+            Assert.Equal(red, controller.DefaultAttribute.BackgroundRgb);
+            Assert.Equal(red, controller.CursorState.Attributes.BackgroundRgb);
+        }
     }
 }
diff --git a/VtNetCore/VirtualTerminal/VirtualTerminalController.cs b/VtNetCore/VirtualTerminal/VirtualTerminalController.cs
index d5213ee..2338836 100644
--- a/VtNetCore/VirtualTerminal/VirtualTerminalController.cs
+++ b/VtNetCore/VirtualTerminal/VirtualTerminalController.cs
@@ -45,6 +45,36 @@
         /// </summary>
         public TerminalAttribute DefaultAttribute { get; }
 
+        public TerminalColor BackgroundColor
+        {
+            get => this.DefaultAttribute.BackgroundRgb;
+            set
+            {
+                this.DefaultAttribute.BackgroundRgb = value;
+                this.NullAttribute.BackgroundRgb = value;
+
+                //
+                // Apply immediately.
+                //
+                SetRgbBackgroundColor((int)value.Red, (int)value.Green, (int)value.Blue);
+            }
+        }
+
+        public TerminalColor ForegroundColor
+        {
+            get => this.DefaultAttribute.ForegroundRgb;
+            set
+            {
+                this.DefaultAttribute.ForegroundRgb = value;
+                this.NullAttribute.ForegroundRgb = value;
+
+                //
+                // Apply immediately.
+                //
+                SetRgbForegroundColor((int)value.Red, (int)value.Green, (int)value.Blue);
+            }
+        }
+
         /// <summary>
         /// Defines the attributes which should be assigned to null character values
         /// </summary>
-- 
2.36.0.windows.1

